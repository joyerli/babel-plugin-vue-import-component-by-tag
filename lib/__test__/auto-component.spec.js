const babel = require('@babel/core');
const _ = require('lodash');
const plugin = require('../index');
const makeStore = require('./__mocks__/store.mock');

jest.mock('vue-record-tags-loader/lib/store', () => require('./__mocks__/store.mock'));

const example = `
import { Button } from 'element-ui';

export default {
  name: 'About',
  /* name: joyer */
  components: {
    [Button.name]: Button,
  },
};
`;

function isElTag(tag) {
  return _.startsWith(tag, 'el-');
}

function isCdTag(tag) {
  return _.startsWith(tag, 'cd-');
}

function isMktTag(tag) {
  return _.startsWith(tag, 'mkt-');
}

it('自动导入', () => {
  makeStore.set('path/to/file.vue', ['section', 'cd-panel', 'el-row', 'el-col', 'mkt-icon',
    'cd-text-action', 'span', 'el-form', 'el-form-item', 'cd-input-money',
    'div', 'cd-button', 'CdButton', 'md-agreement']);

  const { code } = babel.transform(example, {
    filename: 'path/to/file.vue',
    sourceType: 'module',
    plugins: [[plugin, {
      lib(tag) {
        if (isElTag(tag)) {
          return `element-ui/lib/${tag.replace('el-', '')}`;
        }
        if (isCdTag(tag)) {
          return `@candy/ui/lib/component/${tag}`;
        }
        if (isMktTag(tag)) {
          return `comp@/${tag}`;
        }
        return null;
      },
      style(tag) {
        if (isElTag(tag)) {
          const label = tag.replace('el-', '');
          return `element-ui/lib/theme-chalk/${label}.css`;
        }
        return null;
      },
    }]],
  });
  const resetCode = code.replace(/".*babel-plugin-vue-import-component-by-tag/g, '"babel-plugin-vue-import-component-by-tag');
  expect(resetCode).toMatchSnapshot();
});
it('排查指定文件不自动导入', () => {
  makeStore.set('path/to/file.vue', ['section', 'cd-panel', 'el-row', 'el-col', 'mkt-icon',
    'cd-text-action', 'span', 'el-form', 'el-form-item', 'cd-input-money',
    'div', 'cd-button', 'CdButton', 'md-agreement']);

  const { code } = babel.transform(example, {
    filename: '/path/to/file.vue',
    sourceType: 'module',
    plugins: [[plugin, {
      exclude: ['**/to/file.vue'],
      lib(tag) {
        if (isElTag(tag)) {
          return `element-ui/lib/${tag.replace('el-', '')}`;
        }
        if (isCdTag(tag)) {
          return `@candy/ui/lib/component/${tag}`;
        }
        if (isMktTag(tag)) {
          return `comp@/${tag}`;
        }
        return null;
      },
      style(tag) {
        if (isElTag(tag)) {
          const label = tag.replace('el-', '');
          return `element-ui/lib/theme-chalk/${label}.css`;
        }
        return null;
      },
    }]],
  });
  expect(code).toMatchSnapshot();
});
