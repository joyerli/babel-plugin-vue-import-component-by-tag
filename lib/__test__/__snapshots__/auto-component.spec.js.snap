// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`排查指定文件不自动导入 1`] = `
"import { Button } from 'element-ui';
export default {
  name: 'About',

  /* name: joyer */
  components: {
    [Button.name]: Button
  }
};"
`;

exports[`自动导入 1`] = `
"import _default10 from \\"babel-plugin-vue-import-component-by-tag/lib/runtime/register.js\\";
import _default9 from \\"@candy/ui/lib/component/cd-button\\";
import _default8 from \\"@candy/ui/lib/component/cd-input-money\\";
import \\"element-ui/lib/theme-chalk/form-item.css\\";
import _default7 from \\"element-ui/lib/form-item\\";
import \\"element-ui/lib/theme-chalk/form.css\\";
import _default6 from \\"element-ui/lib/form\\";
import _default5 from \\"@candy/ui/lib/component/cd-text-action\\";
import _default4 from \\"comp@/mkt-icon\\";
import \\"element-ui/lib/theme-chalk/col.css\\";
import _default3 from \\"element-ui/lib/col\\";
import \\"element-ui/lib/theme-chalk/row.css\\";
import _default2 from \\"element-ui/lib/row\\";
import _default from \\"@candy/ui/lib/component/cd-panel\\";
import { Button } from 'element-ui';
const _component = {
  name: 'About',

  /* name: joyer */
  components: {
    [Button.name]: Button
  }
};
const components = _component.components || {};

_default10([{
  tag: 'cd-panel',
  component: _default
}, {
  tag: 'el-row',
  component: _default2
}, {
  tag: 'el-col',
  component: _default3
}, {
  tag: 'mkt-icon',
  component: _default4
}, {
  tag: 'cd-text-action',
  component: _default5
}, {
  tag: 'el-form',
  component: _default6
}, {
  tag: 'el-form-item',
  component: _default7
}, {
  tag: 'cd-input-money',
  component: _default8
}, {
  tag: 'cd-button',
  component: _default9
}], components);

_component.components = components;
export default _component;"
`;
